### PR TITLE
Various refactorings towards universe-based regions

### PR DESCRIPTION
**Pull request by nikomatsakis**
These are a variety of refactorings that I did on my lazy normalization. This code is not expected to change the behavior in any particular way, just pave the way for using universes more uniformly in place of the existing "leak check" approach (and, in turn, for normalization).

One of the notable things that this branch does is to adopt the unification table from [the ena project in rust-lang-nursery](https://github.com/rust-lang-nursery/ena/), which is of course branched from the compiler sources.

### CODE REVIEW

File: src/librustc/infer/type_variable.rs [hunk 1]
```diff
@@ -286,90 +300,125 @@ impl<'tcx> TypeVariableTable<'tcx> {
...
+        // the unification table to prefer "lower" vids -- the problem
+        // is that we may have a variable X (from before the snapshot)
+        // and Y (from after the snapshot) which get unified, with Y
+        // chosen as the new root. Now we are "instantiating" Y with a
+        // value, but it escapes into X, but we wouldn't readily see
+        // that. (In fact, earlier revisions of this code had this
+        // bug; it was introduced when we added the `eq_relations`
+        // table, but it's hard to create rust code that triggers it.)
+        //
+        // We could tell the table to prefer lower vids, and then we would
```
Comment (arielb1) [inline src/librustc/infer/type_variable.rs]: We could also store the minimum vid of a node's ~~children~~once-children  on each tree node and propagate that when we unify.

File: src/librustc_typeck/check/_match.rs [hunk 2]
```diff
@@ -285,6 +285,7 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 let element_tys_iter = (0..max_len).map(|_| self.next_ty_var(
```
Comment (arielb1) [inline src/librustc_typeck/check/_match.rs]: This looks ugly - would introduce a `next_global_ty_var` function on fcx.

File: src/librustc/ty/mod.rs [hunk 3]
```diff
@@ -1250,6 +1332,17 @@ pub struct ParamEnv<'tcx> {
     /// want `Reveal::All` -- note that this is always paired with an
     /// empty environment. To get that, use `ParamEnv::reveal()`.
     pub reveal: traits::Reveal,
+
```
Comment (arielb1) [inline src/librustc/ty/mod.rs]: I find `ParamEnv` containing a `universe` rather unsatisfying - it doesn't fit in my mental model.

**Comment by nikomatsakis** [comment]
So @arielb1 and I had [a chat on `#rustc`](https://botbot.me/mozilla/rustc/2017-11-13/?msg=93473964&page=2). We said some interesting stuff. Some notes:

- It'd be nice if we can use universes as a "better leak-check".
- In principle, we could generate a fresh universe (rather than keeping a stack) each time we enter a binder, which would mean that we don't need to track the universe in the parameter environment.
    - Later, we can canonicalize universes anyway, since the only thing that really matters is the relative order in which they were created.
- We do want to be a bit careful with how we integrate universes into caching and so forth anyway.

I have to think over just how I want to adjust this PR. I think I will close it for the time being and mull things over.

**Pull request closed without merging**